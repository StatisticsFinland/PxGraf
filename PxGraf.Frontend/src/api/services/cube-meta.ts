/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeMeta } from "types/cubeMeta";
import { defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for cube metadata result. Cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ICubeMeta} data - The cube meta data as an @see {@link ICubeMeta} object.
 */
export interface ICubeMetaResult {
    isLoading: boolean;
    isError: boolean;
    data: ICubeMeta;
}

const fetchCubeMeta = async(idStack: string[]): Promise<ICubeMeta> => {
    const client = new ApiClient();
    const url = `creation/cube-meta/${idStack.join('/')}`;

    return await client.getAsync(url);
}

export const useCubeMetaQuery = (idStack: string[]): ICubeMetaResult => {
    return useQuery(
        ['cube-meta', idStack],
        () => fetchCubeMeta(idStack),
        defaultQueryOptions
    );
}