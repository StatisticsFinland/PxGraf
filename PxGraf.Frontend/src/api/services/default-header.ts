/* istanbul ignore file */

import ApiClient from "api/client";
import { merge } from "lodash";
import { useQuery } from "react-query";
import { MultiLanguageString } from "types/multiLanguageString";
import { Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for the chart's header result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred while retrieving the header.
 * @property {MultiLanguageString} data - The header data represented as a multi-language string.
 */
export interface IHeaderResult {
    isLoading: boolean;
    isError: boolean;
    data: MultiLanguageString;
}

const fetchDefaultHeader = async(idStack: string[], query: Query): Promise<MultiLanguageString> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        merge(
            { tableReference: buildTableReference(idStack) },
            { variableQueries: query }
        )
    );
  
    const url = 'creation/default-header';
    return await client.postAsync(url, requestBody);
}

export const useDefaultHeaderQuery = (idStack: string[], query: Query): IHeaderResult => {
    return useQuery(
        ['default-header', ...idStack, query],
        () => fetchDefaultHeader(idStack, query),
        {
            ...defaultQueryOptions,
            enabled: query != null,
        }
    );
}