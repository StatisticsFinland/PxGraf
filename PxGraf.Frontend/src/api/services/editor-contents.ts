/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IEditorContentsResponse } from "types/editorContentsResponse";
import { buildCubeQuery, defaultQueryOptions, useDebounceState } from "utils/ApiHelpers";
import { ICubeQuery, Query } from "types/query";

/**
 * Interface for editor contents result. Editor contents are used by the editor to populate the settings for adjusting the visualization.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {IEditorContentsResponse} data - The editor contents as an @see {@link IEditorContentsResponse} object.
 */ 
export interface IEditorContentsResult {
    isLoading: boolean;
    isError: boolean;
    data: IEditorContentsResponse;
}

const fetchEditorContents = async (idStack: string[], query: Query, metaEdits: ICubeQuery): Promise<IEditorContentsResponse> => {
    const client = new ApiClient();
    const url = 'creation/editor-contents';
    const requestBody = JSON.stringify(buildCubeQuery(query, metaEdits, idStack));
    return await client.postAsync(url, requestBody);
}

export const useEditorContentsQuery = (idStack: string[], query: Query, cubeQuery: ICubeQuery): IEditorContentsResult => {
    [idStack, query, cubeQuery] = useDebounceState(1000, idStack, query, cubeQuery);
    const queryKey = ['editor-contents', ...idStack, query, cubeQuery];
    return useQuery(
        queryKey,
        () => fetchEditorContents(idStack, query, cubeQuery),
        {
            ...defaultQueryOptions,
            enabled: query != null,
            keepPreviousData: true,
        }
    );
}