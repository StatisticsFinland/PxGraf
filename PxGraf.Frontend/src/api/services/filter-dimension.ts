/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IValueFilter, Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for variable filter result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {{ [key: string]: string[] }} data - The variable filter data represented by a dictionary. Key is the variable code and value is a list of strings that represent value codes for that variable.
 */
export interface IFilterVariableResult {
    isLoading: boolean;
    isError: boolean;
    data: {[key: string]: string[]}
}

const fetchResolveVariableFilter = async (idStack: string[], variableFilters: { [key: string]: IValueFilter }): Promise<{ [key: string]: string[] }> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        {
            tableReference: buildTableReference(idStack),
            filters: variableFilters,
        }
    );

    const url = 'creation/filter-dimension'
    return await client.postAsync(url, requestBody);
}

export const useResolveVariableFiltersQuery = (idStack: string[], query: Query): IFilterVariableResult => {
    const varQueries = Object.entries(query ?? {});
    const varFilters =  Object.fromEntries(varQueries.map(([variableCode, variableQuery]) => {
        return [variableCode, variableQuery.valueFilter]
    }));

    return useQuery(
        ['filter-dimension', ...idStack, query],
        () => fetchResolveVariableFilter(idStack, varFilters),
        defaultQueryOptions
    );
}