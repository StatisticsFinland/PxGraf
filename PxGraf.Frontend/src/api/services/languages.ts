/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";

const fetchLanguages = async (): Promise<string[]> => {
    const client = new ApiClient();
    const url = "creation/languages/";

    return await client.getAsync(url);
};

/**
 * Parses a list of languages into a string for display.
 * @param {string[]} languages List of languages.
 * @returns All listed languages in one string, in upper case.
 */
export const parseLanguageString = (languages: string[]): string => {
    return `(${languages.join(", ").toUpperCase()})`;
}

/**
 * Chooses a listing language for the given listing item, trying to prefer the selected uiLanguage.
 * @param {string} id Id of the database or table.
 * @param {string[]} languages Languages supported by the listing item.
 * @param {string} uiLanguage Currently selected UI language.
 * @returns UI language, if it is supported by the listing item's languages. Otherwise the default language of the item.
 */
export const getDatabaseListingLanguage = (id: string, languages: string[], uiLanguage: string): string => {
    return languages.includes(uiLanguage) ? uiLanguage : languages[0];
}

export const useLanguagesQuery = (): string[] => {
    const queryResult = useQuery(
        ['languages'],
        () => fetchLanguages(),
        defaultQueryOptions
    );
    return queryResult.data;
};