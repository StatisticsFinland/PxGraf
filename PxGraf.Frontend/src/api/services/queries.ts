/* istanbul ignore file */

import ApiClient from "api/client";
import { useMutation, useQuery } from "react-query";
import { ICubeQuery, Query } from "types/query";
import { IVisualizationSettings } from "types/visualizationSettings";
import { buildCubeQuery, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Response object for saving a query.
 * @property {string} id - The id of the saved query.
 */
export interface ISaveQueryResponse {
    id: string;
}

/**
 * Response object for fetching a saved query.
 * @property query - The query object consisting of the table reference and variable queries @see {@link Query}.
 * @property {IVisualizationSettings} settings - The visualization settings for the query.
 */
export interface IFetchSavedQueryResponse {
    query: {
        tableReference: {
            name: string;
            hierarchy: string[];
        };
        variableQueries: Query;
    } & ICubeQuery;
    settings: IVisualizationSettings;
}

/**
 * Interface for save query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {ISaveQueryResponse} data - The save query response data.
 * @property {function} mutate - Function to mutate the state of the query.
 */
export interface ISaveQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: ISaveQueryResponse;
    mutate: (property: boolean) => void;
}

/**
 * Interface for fetch saved query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {IFetchSavedQueryResponse} data - The fetch saved query response data.
 */
export interface IFetchSavedQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: IFetchSavedQueryResponse;
}

const sendSaveRequest = async (
    idStack: string[],
    query: Query,
    metaEdits: ICubeQuery,
    selectedVisualization: string,
    visualizationSettings: IVisualizationSettings,
    archive: boolean | void
): Promise<ISaveQueryResponse> => {

    const client = new ApiClient();

    const requestBody = JSON.stringify({
        query: buildCubeQuery(query, metaEdits, idStack),
        settings: {
            ...visualizationSettings,
            selectedVisualization: selectedVisualization,
        }
    });
    const url = archive ? 'sq/archive' : 'sq/save';
    return await client.postAsync(url, requestBody);
}

export const fetchSavedQuery = async (queryId: string): Promise<IFetchSavedQueryResponse> => {
    const client = new ApiClient();
    const url = 'sq/' + queryId;
    const response = await client.getAsync(url);

    // Check for VariableQueries and assign to variableQueries
    if (!response.query?.variableQueries && response.query?.VariableQueries) {
        response.query.variableQueries = response.query.VariableQueries;
    }

    return response as IFetchSavedQueryResponse;
};

export const useFetchSavedQuery = (queryId: string): IFetchSavedQueryResult => {
    return useQuery(
        [queryId],
        () => fetchSavedQuery(queryId),
        defaultQueryOptions
    );
};

export const useSaveMutation = (idStack: string[], query: Query, metaEdits: ICubeQuery, selectedVisualization: string, visualizationSettings: IVisualizationSettings): ISaveQueryResult => {
    return useMutation(
        (archive) => sendSaveRequest(idStack, query, metaEdits, selectedVisualization, visualizationSettings, archive),
        { retry: false }
    );
}