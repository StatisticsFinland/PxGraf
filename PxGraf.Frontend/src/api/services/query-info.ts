/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeQuery, IQueryInfo, Query } from "types/query";
import { buildCubeQuery, defaultQueryOptions, useDebounceState } from "utils/ApiHelpers";

/**
 * Interface for query info result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {IQueryInfo} data - The query info data reprented by @see {@link IQueryInfo}.
 */
export interface IQueryInfoResult {
    isLoading: boolean;
    isError: boolean;
    data: IQueryInfo;
}

const fetchQueryInfo = async (idStack: string[], query: Query, metaEdits: ICubeQuery): Promise<IQueryInfo> => {

    const client = new ApiClient();
    const url = 'creation/query-info';
    const requestBody = JSON.stringify(buildCubeQuery(query, metaEdits, idStack));
    return await client.postAsync(url, requestBody);
}

export const useQueryInfoQuery = (idStack: string[], query: Query, cubeQuery: ICubeQuery): IQueryInfoResult => {
    [idStack, query, cubeQuery] = useDebounceState(1000, idStack, query, cubeQuery);
    const queryKey = ['query-info', ...idStack, query, cubeQuery];

    return useQuery(
        queryKey,
        () => fetchQueryInfo(idStack, query, cubeQuery),
        {
            ...defaultQueryOptions,
            enabled: query != null,
            keepPreviousData: true,
        }
    );
}