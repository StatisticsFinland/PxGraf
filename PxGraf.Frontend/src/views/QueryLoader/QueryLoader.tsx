import { fetchSavedQuery } from 'api/services/queries';
import React, { useEffect, useState } from 'react';
import { useParams, useNavigate } from "react-router-dom";
import { Container, CircularProgress, Alert } from '@mui/material';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

const CubeMetaAlert = styled(Alert)`
  margin: 32px;
`;

enum PathProcessingState {
    Idle,
    Loading,
    Error
}

/**
 * Component used for loading a saved query using the query ID from the URL.
 */
export default function QueryLoader() {
    const { t } = useTranslation();
    const sqid = useParams()["*"];
    const [queryFetchError, setQueryFetchError] = useState("");
    const [pathProcessingState, setPathProcessingState] = useState(PathProcessingState.Idle);
    const navigate = useNavigate();

    useEffect(() => {
        setPathProcessingState(PathProcessingState.Loading);
        fetchQueryAndRedirect(sqid);
    }, [sqid]);

    React.useEffect(() => {
        document.title = `${t("pages.sqid")} | PxGraf`;
    }, []);

    const fetchQueryAndRedirect = async (queryId: string) => {
        try {
            const result = await fetchSavedQuery(queryId);
            const url = `/editor/${result.query.tableReference.hierarchy.join('/')}/${result.query.tableReference.name}/`
            navigate(url, { state: { result: result, queryId: queryId } });
        } catch (error) {
            setPathProcessingState(PathProcessingState.Error);
            setQueryFetchError(error.message);
        }
    }

    if (pathProcessingState === PathProcessingState.Loading) {
        return (
            <Container sx={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                <CircularProgress role="loading" />
            </Container>
        );
    }
    else if (pathProcessingState === PathProcessingState.Error) {
        return (
            <Container sx={{ display: 'flex', alignItems: 'flex-start', justifyContent: 'center' }}>
                <CubeMetaAlert severity="error">{t('error.queryLoad') + queryFetchError}</CubeMetaAlert>
            </Container>
        );
    }
}