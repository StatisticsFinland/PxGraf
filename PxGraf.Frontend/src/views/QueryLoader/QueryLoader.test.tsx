import { render, waitFor } from "@testing-library/react";
import QueryLoader from "./QueryLoader";

import { HashRouter } from "react-router-dom";
import { fetchSavedQuery } from 'api/services/queries';

jest.mock('api/services/queries');
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

describe('Rendering test', () => {

    it('renders correctly', () => {
        const { asFragment } = render(
            <HashRouter><QueryLoader /></HashRouter>
        );
        waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });
});

describe('Assertion tests', () => {
    it('displays error state', async () => {
        (fetchSavedQuery as jest.Mock).mockRejectedValueOnce(new Error('fetch error'));
        const { getByRole } = render(<HashRouter><QueryLoader /></HashRouter>);
        await waitFor(() => expect(getByRole('alert')).toBeInTheDocument());
    });

    it('displays loading state', () => {
        (fetchSavedQuery as jest.Mock).mockImplementation(() => new Promise(() => null));
        const { getByRole } = render(<HashRouter><QueryLoader /></HashRouter>);
        expect(getByRole('loading')).toBeInTheDocument();
    });

    it('navigates after successful fetch', async () => {
        (fetchSavedQuery as jest.Mock).mockResolvedValueOnce({
            query: {
                tableReference: {
                    name: 'foo.px',
                    hierarchy: ['bar', 'baz'],
                },
                variableQueries: {},
            },
            settings: {}
        });
        render(<HashRouter><QueryLoader /></HashRouter>);
        await waitFor(() => expect(mockNavigate.mock.calls[0][0]).toEqual("/editor/bar/baz/foo.px/"));
    });
});