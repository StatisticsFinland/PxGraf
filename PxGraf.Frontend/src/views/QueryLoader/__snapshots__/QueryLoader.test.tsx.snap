// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-186hp2m-MuiContainer-root"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
      role="loading"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;
